### PR TITLE
Add several warnings to the GCC build

### PR DESCRIPTION
**Pull request by MeAreJeenius**
Not an exhaustive list by any means. Probably lots of low-hanging fruit left. Definitely lots of other warnings that should be added, but will require a bunch of work.

All of the warnings requiring fixes were pretty trivial fixes, so no refactoring or changes in logic were done here.

### CODE REVIEW

File: src/cpu/core_normal/table_ea.h
Comment (johnnovak) [inline src/cpu/core_normal/table_ea.h:143]: Maybe just use single spaces as separators?

File: src/cpu/core_normal/table_ea.h
Comment (johnnovak) [inline src/cpu/core_normal/table_ea.h:174]: Maybe just use single spaces as separators?

File: src/cpu/modrm.cpp
Comment (johnnovak) [inline src/cpu/modrm.cpp:151]: Maybe just use single spaces as separators?

File: src/cpu/modrm.cpp
Comment (johnnovak) [inline src/cpu/modrm.cpp:175]: Maybe just use single spaces as separators? (same comment for the rest 😄 )

File: src/debug/debug_disasm.cpp
Comment (johnnovak) [inline src/debug/debug_disasm.cpp:316]: Indentation seems to be off with the rest.

File: src/debug/debug_disasm.cpp
Comment (johnnovak) [inline src/debug/debug_disasm.cpp:324]: Indentation seems to be off with the rest.

File: src/debug/debug_disasm.cpp
Comment (johnnovak) [inline src/debug/debug_disasm.cpp:361]: Indentation seems to be off with the rest.

File: src/gui/render_scalers.cpp
Comment (johnnovak) [inline src/gui/render_scalers.cpp:192]: Indentation seems to be off with the rest?

File: src/gui/render_scalers.cpp
Comment (johnnovak) [inline src/gui/render_scalers.cpp:232]: Indentation seems to be off with the rest?

File: include/reelmagic.h
Comment (johnnovak) [inline include/reelmagic.h:49]: Are these unused? If they're internal-only, ideally we should change their names (by convention, static functions are in `lower_snake_case` and they're not prefixed by the module name; e.g. `RENDER_DoStuff` would become `do_stuff`).

**Comment by johnnovak** [comment]
Very nice @MeAreJeenius, that's massive! 🎉 Just a few comments/questions/remarks, otherwise should be good to go once we sort those few things out!

File: include/reelmagic.h
Comment (johnnovak) [reply include/reelmagic.h:49]: Are you sure removing these is ok? Just removed them in `main` and I'm getting compilation errors:

```
FAILED: src/hardware/libhardware.a.p/vga_other.cpp.o
ccache c++ -Isrc/hardware/libhardware.a.p -Isrc/hardware -I../../src/hardware -I../../include -I. -I../.. -Isrc/libs/ghc -I../../src/libs/ghc -Isubprojects/iir1-1.9.3 -I../../subprojects/iir1-1.9.3 -Isrc/libs/loguru -I../../src/libs/loguru -I/opt/local/include/slirp -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/include -I/opt/local/include/libpng16 -I/opt/local/include/SDL2 -fcolor-diagnostics -Wall -Winvalid-pch -Wextra -Wpedantic -std=c++17 -O0 -g -Wno-unknown-pragmas -fpch-preprocess -Weffc++ -Wextra-semi -Wno-deprecated-declarations -fPIC -DLOGURU_STACKTRACES=1 -DLOGURU_USE_LOCALE=0 -D_THREAD_SAFE -MD -MQ src/hardware/libhardware.a.p/vga_other.cpp.o -MF src/hardware/libhardware.a.p/vga_other.cpp.o.d -o src/hardware/libhardware.a.p/vga_other.cpp.o -c ../../src/hardware/vga_other.cpp
../../src/hardware/vga_other.cpp:585:4: error: use of undeclared identifier 'ReelMagic_RENDER_SetPal'
                        RENDER_SetPal(index, r, g, b);
                        ^
../../src/hardware/reelmagic/vga_passthrough.h:26:30: note: expanded from macro 'RENDER_SetPal'
#define RENDER_SetPal        ReelMagic_RENDER_SetPal
                             ^
1 error generated.
[11/19] Compiling C++ object src/hardware/libhardware.a.p/reelmagic_video_mixer.cpp.o
ninja: build stopped: subcommand failed.
```

File: src/cpu/core_normal/table_ea.h
Comment (johnnovak) [reply src/cpu/core_normal/table_ea.h:143]: > Is there any reason I shouldn't just apply clang-format to the whole project except the libs directory in a separate commit?

Yes, don't do that. It can make porting changes over from DOSBox SVN harder than it needs to be.

The "informal" rule of thumb is to reformat only files you're touching, and even then to restrict it to the parts you're touching. And of course ideally always in separate commits.

So we're doing the clang reformatting "lazily", on demand.

File: include/reelmagic.h
Comment (johnnovak) [reply include/reelmagic.h:49]: Hmm, the branch itself compiles fine locally here. Not quite sure what's going on...

File: src/gui/render_scalers.cpp
Comment (johnnovak) [reply src/gui/render_scalers.cpp:232]: This is how the indentation looks in current `main`:

image

After the reformat: 

image

Can you please fix this up so the nice tabular layout is kept? Then enclose the tabular block with `// clang-format off` and `// clang-format on` to exclude it from the Clang formatter? Check out `int10_modes.cpp` for a concrete example.

File: src/cpu/modrm.cpp
Comment (johnnovak) [reply src/cpu/modrm.cpp:151]: Ok these are fine actually.

File: src/debug/debug_disasm.cpp
Comment (johnnovak) [reply src/debug/debug_disasm.cpp:316]: This looked like this before:

image

Now it looks like this; the tabular formatting is lost:

image

**Comment by johnnovak** [comment]
@MeAreJeenius I went through some of the formatting issues I commented on you marked as resolved, and posted screenshot examples of how the clang formatter ruined the nice tabular format.

I appreciate your proactive work on getting rid of the warnings, but removing these well-crafted tabular layouts and leaving restoring the layout to the next person who's gonna touch these files in the future is a no-go. We should not leave the files in a worse state after touching them, and the tabular layout is important, it makes things more apparent and easier to parse visually.

Can you please fix up the layouts and restore them to their former glory? 😄  After that, please fence them off with `// clang-format off` and `// clang-format on` (see `int10_modes.cpp` for a concrete example). Thank you!

**Comment by johnnovak** [comment]
> @johnnovak I marked them resolved to make the page easier to read through. It's a lot of repetition, and I've already fixed them locally.

Ah, right, thanks for clarifying that 😅 Yeah we don't really have an agreed upon standard way of resolving comments, so I kinda thought you were just dismissing them. Apologies 😄

File: src/cpu/core_normal/table_ea.h
Comment (LowLevelMahn) [reply src/cpu/core_normal/table_ea.h:143]: there is a script that formats only the "changed lines" of the last commit - see: script/format-commit.sh

works for me on Windows only in a MSYS2 environment :/

1. make you're changes
2. commit
3. use the script/format-commit.sh - every changed line gets formatted
(you still need to keep out these tables johnnovak is talking about)
4. rebase or squash the formatted changes back into in one commit

Comment (johnnovak) [reply src/cpu/core_normal/table_ea.h:143]: That's cool, I never tried it myself. I use a Vim plugin for Clang Format and pretty much reformat what I've changed manually 100% of the time.

File: include/reelmagic.h
Comment (johnnovak) [reply include/reelmagic.h:49]: I think @kcgen did not want to change the code much so it's easier to merge further improvements. But my understanding is that the original author will update the version in our repo (if he'll ever get to it). I vaguely remember some correspondence between @kcgen and the author about that. So probably it's okay to make a few judicious improvements to it (and I had to touch it too a bit because of my recent image capture / video related refactorings & improvements).

File: src/cpu/core_normal/table_ea.h
Comment (LowLevelMahn) [reply src/cpu/core_normal/table_ea.h:143]: i was forced to use it - for my micro pulls :) - nothing compared to your output

**Comment by johnnovak** [comment]
> I changed how the warnings are set in meson. Now they are in extra_args, which filters out unavailable warnings for the current compiler, so different compilers are no longer an issue.
> 
> To compensate, I had to disable warnings altogether on some third-party code in src/libs. Not ideal, but meson really doesn't want to make this easy. Definitely open to suggestions if there is a nicer way.

I'm really a beginner with Meson, so I defer to the others' opinion on this. @shermp if you think it's fine, then it's fine by me too 😄

**Comment by johnnovak** [comment]
This is a heroic cleanup job @MeAreJeenius , congrats 🥇 Formatting of tabular stuff looks now good, thanks for addressing that!

Would these changes allow us to lower our allowed warning limits? We've done that periodically, e.g. check out these commits 49a80c3b373c0e2,  9a374d88bc5b

If not, that's fine too, just checking!

**Comment by johnnovak** [comment]
How are you going with this one @MeAreJeenius ?

File: .github/workflows/linux.yml
Comment (johnnovak) [inline .github/workflows/linux.yml:99]: That's a bit unfortunate to have to increase the max warnings. Going from 2 to 21 is quite significant. Ideally, we want to go from 2 to 0 😄 Can't you find a better workaround so we can keep it at 2?

File: src/hardware/gameblaster.h
Comment (johnnovak) [inline src/hardware/gameblaster.h:39]: I'm unsure about the value and necessity of this change. Can you explain why this is an improvement in your view?

This blurs the distinction between system-provided and bundled libraries; I actually _prefer_ to make that distinction explicit.

Sure, if you move files around, you need to readjust the relative paths. But how often do we do that? Almost never, and even then it's a 2 min job to adjust the paths.

File: include/reelmagic.h
Comment (johnnovak) [inline include/reelmagic.h:41]: You're changing the types here. Would be better to separate the type changes out into their own commit?

File: include/reelmagic.h
Comment (johnnovak) [inline include/reelmagic.h:44]: Same here. Not sure why you needed to change it back from uint8_t. I have changed bpp to uint8_t in most places and I'm finishing the job in my big VGA related PR currently under review.

File: src/hardware/vga_dac.cpp
Comment (johnnovak) [inline src/hardware/vga_dac.cpp:25]: Ok so this commit is doing a lot more than just removing a few unused declarations. You could either call it "Various cleanup changes" but better yet, you could split the unrelated changes, which should be easy to do.

File: src/hardware/vga_draw.cpp
Comment (johnnovak) [inline src/hardware/vga_draw.cpp:976]: The ReelMagic related cleanups/renames most definitely warrant their own commit.

File: tests/rgb_tests.cpp
Comment (johnnovak) [inline tests/rgb_tests.cpp:25]: Making changes to code should be in a separate commit after this. This commit should only do what the commit message says.

File: .github/workflows/linux.yml
Comment (johnnovak) [reply .github/workflows/linux.yml:99]: I did read the commit description, but I'm wondering if there's a way to keep `max_warnings` at 2 while keeping your change 😄 

E.g. if you can use some pragma or "magic comment" to make these problematic compilers happy to ignore these 20 warnings, that's a much better workaround than globally increasing the number of allowed warnings for all compilers from 2 to 21 wholesale. That's a bit too heavy-handed approach, and I'd actually prefer not to do this change at all and keep the `max_warnings` level at 2.

File: src/hardware/gameblaster.h
Comment (johnnovak) [reply src/hardware/gameblaster.h:39]: Good points, I agree that not having to use relative paths is nicer. My problem was mainly about the angled bracket vs double-quote convention. Ultimately, I want to keep this convention:

- Angle brackets are used for standard C and C++ libraries and 3rd party dependencies not embedded into the project (e.g. `#include <SDL.h>`) 
- Double-quotes are used for everything else, basically files that live inside the repo. That includes 3rd party dependencies embedded into the repository.

If switching to double-quotes would still work, I'm all for this!

File: include/reelmagic.h
Comment (johnnovak) [reply include/reelmagic.h:41]: Sure, but separate them out into their own commit please? They don't belong the the commit they're under now.

File: include/reelmagic.h
Comment (johnnovak) [reply include/reelmagic.h:44]: Separate commit please. I'm gonna change these back to uint8_t when my PR gets merged, but whatever.

File: src/hardware/vga_dac.cpp
Comment (johnnovak) [reply src/hardware/vga_dac.cpp:25]: Well, the commit's title must reflect what the commit does 😄 So either change the commit message so it's a lot more broad (worse solution), or split the commit and make the commit messages of the resulting commits more specifi (better solution) 😄

File: .github/workflows/linux.yml
Comment (johnnovak) [reply .github/workflows/linux.yml:99]: Ok kinda missed that, my bad. I thought it's for all Linux builds. If others are happy with this, I think it's acceptable compromise perhaps. @FeralChild64 @shermp ?

File: tests/rgb_tests.cpp
Comment (johnnovak) [inline tests/rgb_tests.cpp:123]: I saw you've addressed most of my comments, thanks for that. Not sure if you were going to address this later or have just missed it. So updating `rgb_tests.cpp` should split out into its own commit too from the "Add several warnings to the GCC build" commit. Apologies if this is already on your TODO list 😅

**Comment by johnnovak** [comment]
@MeAreJeenius is this ready for final review now then? Can you please also rebase to `main`?

**Review by johnnovak — APPROVED** [review]
> Should be good to go @johnnovak .

Cheers @MeAreJeenius, that was a massive amount of work! 🎉 Thanks for bearing with me and addressing all the comments!

**Pull request merged**
